batch-jberet/src/main/java/org/wildfly/extension/batch/jberet/deployment/BatchEnvironmentProcessor.java

```diff
@@ -135,11 +134,12 @@ public void deploy(final DeploymentPhaseContext phaseContext) throws DeploymentU
             final ServiceBuilder<ArtifactFactory> artifactFactoryServiceBuilder = serviceTarget.addService(artifactFactoryServiceName, artifactFactoryService);
 
             // Register the bean manager if this is a CDI deployment
-            if (WeldDeploymentMarker.isPartOfWeldDeployment(deploymentUnit)) {
-                BatchLogger.LOGGER.tracef("Adding BeanManager service dependency for deployment %s", deploymentUnit.getName());
-                artifactFactoryServiceBuilder.addDependency(BatchServiceNames.beanManagerServiceName(deploymentUnit), BeanManager.class,
-                        artifactFactoryService.getBeanManagerInjector());
-            }
+            support.getOptionalCapabilityRuntimeAPI(WELD_CAPABILITY_NAME, WeldCapability.class).ifPresent(api -> {
```

Comment:
The capability API is not great here. The use of the WeldCapability class is not guarded by anything that ensures the class doesn't have to be loaded if the cap is not present. CapababilityServiceSupport.hasCapability is meant to be that guard. Using that combined with an Optional.ifPresent amounts to a double-check of whether the capability is there.  Perhaps I should have just made NoSuchCapabilityException an unchecked exception. Making it checked basically helps force coders to think about the possibility of it not being there (i.e. call hasCapability first) but having to catch a checked exception is too big a cost.

Reply:
Actually what you do at https://github.com/wildfly/wildfly/pull/11802/files#diff-b0185cd463b77ad4a276681302a1d18eR123 is reasonable enough looking, so I think the odd thing here would be the use of ifPresent.  Which isn't truly necessary.  You wouldn't even have to store a ref to WeldCapability in a var if addBeanManagerService did the isPartOfWeldDeployment check itself.

Reply:
> The capability API is not great here. The use of the WeldCapability class is not guarded by anything that ensures the class doesn't have to be loaded if the cap is not present. 

I would say it is warded, the `ifPresent` only will unwrap the optional value, exposing the WeldCapability class to the caller, if only if the optional is not empty. If it is empty, then the code will not use a `WeldCapability` instance, and then the calling code could have the module that loads this class as an optional dependency. 
The optional will be empty if only if the capability is not present. Indeed `support.hasCapability(WELD_CAPABILITY_NAME)` returning false should be equivalent. If they are not equivalent, a bug is present in any of them. Both methods will end up executing the same code: CapabilityRegistry.findSatisfactoryCapability

One possible problem with the `Optional.ifPresent` is that it requires a consumer as an argument, and any variable initialized outside of the consumer lamda expression must be final or effectively final to be used in it. If you need to use a variable that is not final or effectively final, then the caller is forced to use `support.hasCapability(WELD_CAPABILITY_NAME)` or `Optional.isPresent()` check before using the API to avoid the lamda expression, that means there could be inconsistencies in the code. @bstansberry If you still think this is a good argument to remove the uses of `ifPresent`, then I can change the PR according to use always `support.hasCapability(WELD_CAPABILITY_NAME)` check before getting the API, keeping all the code using the same structure. What do you think?

> Perhaps I should have just made NoSuchCapabilityException an unchecked exception. Making it checked basically helps force coders to think about the possibility of it not being there (i.e. call hasCapability first) but having to catch a checked exception is too big a cost.

I think it is fine, we now should use the method with the checked exception if the capability is mandatory to the caller, and then catch and react according if the capability is not there. I think if we use an optional, the caller is still forced to think what to do if the capability is not present, although this method should now be used if the capability is optional in the system. Yes, the try/catch of the exception has a cost and the virtual machine is unable to do some optimizations in those blocks.

The inclusion of `isPartOfWeldDeployment` inside of `addBeanManagerService` or `registerExtensionInstance` makes sense; both methods only are relevant if the Deployment Unit is part of a Weld deployment, although we could lose some flexibility in the API, for example, to do something before `addBeanManagerService` or `registerExtensionInstance` and after checking that the deployment pass `isPartOfWeldDeployment` without a double execution of `isPartOfWeldDeployment,` see https://github.com/wildfly/wildfly/pull/11802/files#diff-9c678c3939366680a48d047f02f37d41R139 or https://github.com/wildfly/wildfly/pull/11802/files#diff-cdf1bb068d2ed81a62e021e1e0419c2aR1156. 
The effect is a double execution of `isPartOfWeldDeployment,` but it is a very cheap operation so we can do it. @bstansberry Do you think we can assume the loss of this little flexibility?

Reply:
Re: the use of WeldCapability being warded, the issue isn't the Optional API use, it's this:

```
support.getOptionalCapabilityRuntimeAPI(WELD_CAPABILITY_NAME, WeldCapability.class)
```

That has no guard.

If there is a case where the isPartOfWeldDeployment check ends up happening twice, that's fine. It's not particularly expensive.

If you prefer staying with the checked exception, that's fine with me. I originally wrote the current approach so I thought it was right then, and have _minor_ doubts now. So my opinion on changing it is not a strong one. :) And your Optional addition shows there's a reasonable way to avoid the exception.

Reply:
@bstansberry My apologies, I was completly blind on this, yes, WeldCapability.class is a reference the optional dependency on wildfly-weld-common.

I prefer the optional version, it removes bolerplate code in the client if the capability is optional.